### PR TITLE
StridedSlice : slice and reverse new ref implementations

### PR DESCRIPTION
**Pull request by pelszkow**
Ticket: 37484

### CODE REVIEW

**Review by tomdol — APPROVED** [review]

**Comment by jdanieck** [comment]
@pelszkow please update the ticket with performance measurements in comparison with old version. Also do the same regarding ngraph library size.

@ilyachur please have a look at this new approach. Intention is to use these new iterators also in other reference implementations.

**Review by jdanieck — APPROVED** [review]

**Review by vladislav-volkov — COMMENTED** [review]
I calculated algorithm complexity, and it's equal to Dimension1 * Dimension2 * ... * DimensionN * Rank.
The Rank multiplier appears due to the coordinate_index() function.
I think we can avoid using the coordinate_index() and reduce the complexity of the algorithm.
A simplified pseudocode for StridedSlice might look like this.

```
int wpos = 0;   // write position  
int rpos = coordinate_index(first_coordinate);   // read position. We calculate first coordinate index.
// Calculate row strides
int row_strides[dim];
calc_row_major_strides(row_strides, src_shape);
/*0*/ for(int i = 0; i < dim0; i += abs(stride0))
      {
          rpos += (stride0 - 1) * row_strides[0];
/*1*/     for(int j = 0; j < dim1; j += abs(stride1))
          {
              rpos += (stride1 - 1) * row_strides[1];
/*2*/         for(int k = 0; k < dim2; k += abs(stride2))
              {
                  rpos += (stride2 - 1) * row_strides[2];
/*3*/             for(int l = 0; l < dim3; l += abs(stride3))
                  {
                      rpos += (stride3 - 1) * row_strides[3];
                      for(int m = 0; m < dim4; m += abs(stride4))
                      {
                          rpos += stride4;
                          dst[wpos++] = dst[rpos];
                      }
                  }
              }
          }
      }
 ```     
I didn't use lower bound and upper bound for loops ranges for simplification.
Also I unrolled nested loops for coordinates iterator.
Here strideN can be negative or positive.
I would like to mention that inner loop can be vectorised by compiler because we don't do any checks and jumps.
Can we implement something like this?

**Comment by pelszkow** [comment]
@slyalin, @ilyachur, @lazarevevgeny , @jane-intel , pleas review this PR.

**Review by tomdol — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/coordinate_index.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#pragma once
+
+#include <cstddef>
+
+namespace ngraph // FWD
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/coordinate_index.hpp:20]: nit: why do we need this comment?

File: ngraph/core/reference/include/ngraph/coordinate_index.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+    class Coordinate;
+    class Shape;
+} // namespace ngraph
+
+namespace ngraph
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/coordinate_index.hpp:26]: Why do you declare the `ngraph` namespace 2 times?

File: ngraph/core/reference/include/ngraph/coordinate_range.hpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+#include "ngraph/coordinate.hpp"
+#include "ngraph/shape.hpp"
+#include "ngraph/strides.hpp"
+
+namespace ngraph
+{
+    namespace coordinates
+    {
+        template <typename Range>
+        class RangeIterator
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/coordinate_range.hpp:31]: Could you add a documentation? Why do we need these templates and etc.?

File: ngraph/core/reference/include/ngraph/coordinate_transform.hpp [hunk 4]
```diff
@@ -31,11 +31,18 @@ namespace ngraph
     ///             {1,0}, {1,1}, {2,2}
     class CoordinateIterator
     {
-    public:
         /// \brief Coordinates iterator constructor
         /// \param target_shape The target shape for coordinates iteration
         /// \param is_end The flag indicates that the coordinate iterator is the last.
-        CoordinateIterator(const Shape& target_shape, bool is_end = false);
+        CoordinateIterator(const Shape& target_shape, bool is_end);
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/coordinate_transform.hpp:37]: You moved this constructor to private. Do we still have cases where `is_end` is true?

**Comment by ilyachur** [comment]
Do you have any performance numbers before and after?

**Comment by ilyachur** [comment]
@pelszkow Did you discuss with @vladislav-volkov his proposal? What do you think about it?

File: ngraph/core/reference/include/ngraph/coordinate_index.hpp (hunk 1 above)
Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/coordinate_index.hpp:20]: Is not required. I will remove it.

File: ngraph/core/reference/include/ngraph/coordinate_index.hpp (hunk 2 above)
Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/coordinate_index.hpp:26]: My intention was to split stuffs which are introduce from includes/FWD. 
@ilyachur, should I join FWD with function declaration?

File: ngraph/core/reference/include/ngraph/coordinate_range.hpp (hunk 3 above)
Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/coordinate_range.hpp:31]: Sure.

File: ngraph/core/reference/include/ngraph/coordinate_transform.hpp (hunk 4 above)
Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/coordinate_transform.hpp:37]: Yes we have: https://github.com/openvinotoolkit/openvino/blob/2b70fa14732d3d5740b4525ad38a6877f2f86959/ngraph/core/reference/src/coordinate_transform.cpp#L537

**Comment by pelszkow** [comment]
> Do you have any performance numbers before and after?

Yes, performance result are in the ticket. Should I put it here to?

**Comment by pelszkow** [comment]
> @pelszkow Did you discuss with @vladislav-volkov his proposal? What do you think about it?

@vladislav-volkov suggest another way to implement this op.

**Comment by jdanieck** [comment]
> > @pelszkow Did you discuss with @vladislav-volkov his proposal? What do you think about it?
> 
> @vladislav-volkov suggest another way to implement this op.

So the real question here is do we need to once again lower abstraction level as Vladislav suggested (don't use coordinate_range) which possibly can give us further performance improvements or can we settle on what was achieved already and have a cleaner code. If we choose to use coordinate_range it could be used also in other ref impls and simplify its implementation.

**Comment by vladislav-volkov** [comment]
> > > @pelszkow Did you discuss with @vladislav-volkov his proposal? What do you think about it?
> > 
> > 
> > @vladislav-volkov suggest another way to implement this op.
> 
> So the real question here is do we need to once again lower abstraction level as Vladislav suggested (don't use coordinate_range) which possibly can give us further performance improvements or can we settle on what was achieved already and have a cleaner code. If we choose to use coordinate_range it could be used also in other ref impls and simplify its implementation.

Can we implement SliceRange, which encapsulates the algorithm I suggested? So we can use it elsewhere in the future.

**Comment by pelszkow** [comment]
> > > > @pelszkow Did you discuss with @vladislav-volkov his proposal? What do you think about it?
> > > 
> > > 
> > > @vladislav-volkov suggest another way to implement this op.
> > 
> > 
> > So the real question here is do we need to once again lower abstraction level as Vladislav suggested (don't use coordinate_range) which possibly can give us further performance improvements or can we settle on what was achieved already and have a cleaner code. If we choose to use coordinate_range it could be used also in other ref impls and simplify its implementation.
> 
> Can we implement SliceRange, which encapsulates the algorithm I suggested? So we can use it elsewhere in the future.
> Also I would rename SliceRange to SliceIterator because main goal of this class is iteration over tensor coordinates subset.

I'm afraid we can't bind @vladislav-volkov approach with `SliceRange`. `SliceRange::Iterator` by design produce index by index for upper layer operation and @vladislav-volkov proposition is doing assignment inside most nested for-loop. `SliceRange` might be optimize by dims-stride precalculations to reduce computation complexity in `SliceRange::increment` but then we reduce code clarity and higher level of abstraction. 

p.s.:
Profiler output after `op::v0::Concat` boost show that `std::make_share` now takes over 7% of `LoadNetwork`
```
Node::constant_fold                   : 51.0%
-> op::v1::StridedScelid::evalueate   : 19.0%
-> op::v1::VariadicSplit::evalueate   : 13.6%
-> std::make_shared>op::v0::Constant> : 7.5%
-> op::v0::Concat::evaluate           : 6.1%
-> op::v1::Add::evaluate              : 3.4%
```

**Comment by vladislav-volkov** [comment]
> I'm afraid we can't bind @vladislav-volkov approach with `SliceRange`. `SliceRange::Iterator` by design produce index by index for upper layer operation and @vladislav-volkov proposition is doing assignment inside most nested for-loop.

Why can't we clone the CoordinateIterator class with an extra field that will hold the index? We can update this index on each iteration. In CoordinateUterator are missing only two things:
1. Strides for iteration
2. lowe and upper bound for each dimension

**Comment by slyalin** [comment]
> > I'm afraid we can't bind @vladislav-volkov approach with `SliceRange`. `SliceRange::Iterator` by design produce index by index for upper layer operation and @vladislav-volkov proposition is doing assignment inside most nested for-loop.
> 
> Why can't we clone the CoordinateIterator class with an extra field that will hold the index? We can update this index on each iteration. In CoordinateIterator are missing only two things:
> 
> 1. Strides for iteration
> 2. lower and upper bound for each dimension

@pelszkow, could you try out this approach proposed by @vladislav-volkov or somehow estimate pros/cons in more formal way? We are interested in more optimized references as there is a plan to reuse them as the fallback kernels implementation inside CPU plugin. But in the same time, the code clarity is also important. From the conversation above it is not clear how much (if any) clarity we are going to loose and how big performance gain we are going to achieve by following the approach.

**Comment by ilyachur** [comment]
> > > I'm afraid we can't bind @vladislav-volkov approach with `SliceRange`. `SliceRange::Iterator` by design produce index by index for upper layer operation and @vladislav-volkov proposition is doing assignment inside most nested for-loop.
> > 
> > 
> > Why can't we clone the CoordinateIterator class with an extra field that will hold the index? We can update this index on each iteration. In CoordinateIterator are missing only two things:
> > 
> > 1. Strides for iteration
> > 2. lower and upper bound for each dimension
> 
> @pelszkow, could you try out this approach proposed by @vladislav-volkov or somehow estimate pros/cons in more formal way? We are interested in more optimized references as there is a plan to reuse them as the fallback kernels implementation inside CPU plugin. But in the same time, the code clarity is also important. From the conversation above it is not clear how much (if any) clarity we are going to loose and how big performance gain we are going to achieve by following the approach.

I agree with Sergey. Can we try this approach? Looks like this approach can allow to get additional speed up.

File: ngraph/core/reference/include/ngraph/coordinate_index.hpp (hunk 2 above)
Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/coordinate_index.hpp:26]: Ok, I got your point.
I prefer to have one namespace but it is my subjective opinion and I am ok with current approach.

Comment (pelszkow) [reply ngraph/core/reference/include/ngraph/coordinate_index.hpp]: File removed.

**Comment by pelszkow** [comment]
Brand new slice and reverse implementations.

**Review by tomdol — APPROVED** [review]

**Comment by vladislav-volkov** [comment]
Could you please write algorithm without recursion?
Also I see a lot of similar code which used for tensor coordinates iteration. I propose to extract it to a separate class.

**Review by vladislav-volkov — COMMENTED** [review]
Could you please write algorithm without recursion? We can crash with stack overflow error for tensors with big dimensions. I suppose that the recursive function call for each iteration is too expensive.
Also I see a lot of similar code which used for tensor coordinates iteration. I propose to extract it to a separate class.

**Comment by jdanieck** [comment]
@pelszkow please adhere to [review good practices](https://github.com/openvinotoolkit/openvino/wiki/Contribute#pr-review-good-practices) and do not force-push.

**Review by ilyachur — APPROVED** [review]

**Comment by pelszkow** [comment]
@vladislav-volkov , can you check if this implementation is fine?

**Review by vladislav-volkov — APPROVED** [review]

**Comment by pelszkow** [comment]
@ilyachur, PR ready to be merge.

**Pull request merged**